src/Microsoft.AspNet.OData.Shared/EdmChangedObjectCollectionOfT.cs

```diff
@@ -0,0 +1,297 @@
+                        if(contentId != null)
+                        {
+                            edmDeletedObject.Id = contentId.ToString();                            
+                        }
+                        else
+                        {
+                            edmDeletedObject.Id = string.Empty;
+                        }
+
+                        CopyInstanceAnnotations(changedObj, edmDeletedObject);
+                        edmDeletedObject.TransientInstanceAnnotationContainer.AddResourceAnnotation("Core.DataModificationException", dataModificationExceptionType);                       
+                        edmChangedObject = edmDeletedObject;
+                        break;
+                    }
+                case DataModificationOperationKind.Delete:
+                case DataModificationOperationKind.Unlink:
+                    {
+                        Type type = typeof(EdmDeltaEntityObject<>).MakeGenericType(typeof(TStructuralType));
+                        EdmDeltaEntityObject edmDeltaEntityObject = Activator.CreateInstance(type, EntityType) as EdmDeltaEntityObject;
+                        foreach(PropertyInfo property in typeof(TStructuralType).GetProperties())
+                        {
+                            edmDeltaEntityObject.TrySetPropertyValue(property.Name, property.GetValue(originalObj));
+                        }
+
+                        CopyInstanceAnnotations(changedObj, edmDeltaEntityObject);
+                        edmDeltaEntityObject.TransientInstanceAnnotationContainer.AddResourceAnnotation("Core.DataModificationException", dataModificationExceptionType);
+                        edmChangedObject = edmDeltaEntityObject;
+                        break;
+                    }
+            }
+        }
+
+        private static void CopyInstanceAnnotations(dynamic changedObj, IEntityObjectInstanceAnnotations destinationObject)
+        {
+            //This is for copying both Persistent and Transient Instance Annotations.
+            IEntityObjectInstanceAnnotations entityObject = changedObj as IEntityObjectInstanceAnnotations;
+
+            if (entityObject != null)
+            {
+                destinationObject.PersistentInstanceAnnotationsContainer = entityObject.PersistentInstanceAnnotationsContainer;
+                destinationObject.TransientInstanceAnnotationContainer = entityObject.TransientInstanceAnnotationContainer; 
+            }           
+        }
+
+        private static TStructuralType GetFilteredItem(Type type, IEnumerable<IEdmStructuralProperty> keys, IEnumerable<TStructuralType> originalList, IEdmChangedObject changedObject)
+        {   
+            //This logic is for filtering the object based on the set of keys,
+            //There will only be very few key elements usually, mostly 1, so performance wont be impacted.
```

Comment:
this logic is very expensive for a large list as it uses LINQ to search the entire list in case there are more than one match (which, in the common case, there would not be, but LINQ doesn't know that).  Better would be if you got the key values up front and then walked through originalList comparing the keys and breaking if/when you found a match to all keys.

For example, the following code is more verbose but much more efficient for large lists and/or multiple key values (especially because it moves the type.GetProperty out of the loop) (note: it assume you call ToArray() on Entity.Keys() on line 138, which is probably a good idea anyway since, although there may only be one key, finding that key can be expensive if not cached):

~~~c#
        private static TStructuralType GetFilteredItem(Type type, IEdmStructuralProperty[] keys, IEnumerable<TStructuralType> originalList, IEdmChangedObject changedObject)
        {
            // This logic is for filtering the object based on the set of keys

            // First get the key values for the changed object
            object keyValue;
            object[] keyValues = new object[keys.Length];
            PropertyInfo[] propertyInfos = new PropertyInfo[keys.Length];
            for (int i = 0; i < keys.Length; i++)
            {
                string keyName = keys[i].Name
                changedObject.TryGetPropertyValue(keyName, out keyValue);
                keyValues[i] = keyValue;
                propertyInfos[i] = type.GetProperty(keyName);
            }

            // Now iterate through the originalList until we find the item with the matching key values
            foreach (TStructuralType item in originalList)
            {
                for (int i = 0; i < keyValues.Length; i++)
                {
                    if (propertyInfos[i].GetValue(item) != keyValues[i])
                    {
                        // Not a match, so try the next one
                        continue;
                    }
                }

                return item;
            }

            return default(TStructuralType);
        }
~~~